libpromises/eval_context.c

```diff
@@ -2922,15 +2922,21 @@ void cfPS(EvalContext *ctx, LogLevel level, PromiseResult status, const Promise
     assert(pp != NULL);
     assert(attr != NULL);
 
-    if (level >= LOG_LEVEL_VERBOSE)
+    /* Either logging something (based on 'fmt') or logging nothing. */
+    assert(((fmt != NULL) && (*fmt != '\0')) || (level == LOG_LEVEL_NOTHING));
+
+    if ((fmt != NULL) && (*fmt != '\0'))
     {
-        LogPromiseContext(ctx, pp);
-    }
+        if (level >= LOG_LEVEL_VERBOSE)
+        {
+            LogPromiseContext(ctx, pp);
+        }
 
-    va_list ap;
-    va_start(ap, fmt);
-    VLog(level, fmt, ap);
-    va_end(ap);
+        va_list ap;
+        va_start(ap, fmt);
+        VLog(level, fmt, ap);
+        va_end(ap);
+    }
```

Comment:
I'd prefer an if-else, using the `NULL_OR_EMPTY` macro, and not checking fmt twice:

```C
if (!NULL_OR_EMPTY(fmt)) // fmt has some log message
{
    if (level >= LOG_LEVEL_VERBOSE)
    {
        LogPromiseContext(ctx, pp);
    }
    va_list ap;
    va_start(ap, fmt);
    VLog(level, fmt, ap);
    va_end(ap);
}
else // No log message in fmt
{
    assert(level == LOG_LEVEL_NOTHING);
    // Nothing to log
}
```

No big deal, of course.